pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/BlockingMultiStreamConsumer.java

```diff
@@ -0,0 +1,246 @@
+  }
+
+  /**
+   * Reads the next block for any ready mailbox or blocks until some of them is ready.
+   *
+   * The method implements a sequential read semantic. Meaning that:
+   * <ol>
+   *   <li>EOS is only returned when all mailboxes already emitted EOS or there are no mailboxes</li>
+   *   <li>If an error is read from a mailbox, the error is returned</li>
+   *   <li>If data is read from a mailbox, that data block is returned</li>
+   *   <li>If no mailbox is ready, the calling thread is blocked</li>
+   * </ol>
+   *
+   * Right now the implementation tries to be fair. If one call returned the block from mailbox {@code i}, then next
+   * call will look for mailbox {@code i+1}, {@code i+2}... in a circular manner.
+   *
+   * In order to unblock a thread blocked here, {@link #onData()} should be called.   *
+   */
+  public E readBlockBlocking() {
+    if (LOGGER.isTraceEnabled()) {
+      LOGGER.trace("==[RECEIVE]== Enter getNextBlock from: " + _id + " mailboxSize: " + _mailboxes.size());
+    }
+    // Standard optimistic execution. First we try to read without acquiring the lock.
+    E block = readDroppingSuccessEos();
+    long timeoutMs = _deadlineMs - System.currentTimeMillis();
+    boolean timeout = false;
+    try {
+      while (block == null) { // we didn't find a mailbox ready to read, so we need to be pessimistic
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug("==[RECEIVE]== Blocked on : " + _id + ". " + System.identityHashCode(_newDataReady));
+        }
+        timeout = _newDataReady.poll(timeoutMs, TimeUnit.MILLISECONDS) == null;
+        LOGGER.debug("==[RECEIVE]== More data available. Trying to read again");
+        block = readDroppingSuccessEos();
+      }
+
+      if (timeout) {
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.warn("==[RECEIVE]== Timeout on: " + _id);
+        }
+        _errorBlock = onTimeout();
+        return _errorBlock;
+      } else if (LOGGER.isTraceEnabled()) {
+        LOGGER.trace("==[RECEIVE]== Ready to emit on: " + _id);
+      }
+    } catch (InterruptedException ex) {
+      return onException(ex);
+    }
+    return block;
+  }
+
+  /**
+   * This is a utility method that reads tries to read from the different mailboxes in a circular manner.
+   *
+   * The method is a bit more complex than expected because ir order to simplify {@link #readBlockBlocking} we added
+   * some extra logic here. For example, this method checks for timeouts, add some logs, releases mailboxes that emitted
+   * EOS and in case an error block is found, stores it.
+   *
+   * @return the new block to consume or null if none is found. EOS is only emitted when all mailboxes already emitted
+   * EOS.
+   */
+  @Nullable
+  private E readDroppingSuccessEos() {
+    if (System.currentTimeMillis() > _deadlineMs) {
+      _errorBlock = onTimeout();
+      return _errorBlock;
+    }
+
+    E block = readBlockOrNull();
+    while (block != null && isEos(block) && !_mailboxes.isEmpty()) {
```

Comment:
(minor) Can `_mailboxes` be empty if `block != null`?

Reply:
Yep, when we read the last EOS from the last mailbox.

Reply:
Why? We remove mailbox within the loop

Reply:
You are right. We only return EOS if there is an EOS in some mailbox, which implies there has to be at least one mailbox